Feature/init dashboards metrics

lennarkivimae opened the pull request.

Commit:
feat:auto analytics adder workflow

Adding interval to add analytics panel for all of the dashboards. Eases
maintenance in team environment, where dev-s don't need to think about
adding an panel manually. Interval runs once in 24 hours

Commit:
auto-analytics-adder - separate general api methods from worker

Commit:
auto-analytics-adder - increase readability, optimizations

Restructure input order
Add extra params to increase testability / flexibility

Commit:
auto-analytics-adder - remove invalid comment

Commit:
auto-analytics-adder - fix issues with kong usage

Add default value for timeout, for easier usage

Commit:
init-dashboard-metrics - move GetDashboards to api.go

GetDashboards is going to be used outside of worker, makes sense to bundle it with api

Commit:
init-dashboard-metrics - add google uuid package to generate uuid for metrics

Commit:
init-dashboard-metrics - Make structs for nested objects in Payload

To generate initial payload, we need to create similar payload data. This enables us to create each struct

Commit:
init-dashboard-metrics - Export process payload

Use already created process to generate initial metrics

Commit:
init-dashboard-metrics - implement metrics initializer for existing dashboards

Creating zeroed metrics helps to track unused dashboards

atsu85 commented on server/initializer/initializer.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+func createDashboardPayload(dashboardUid string, dashboardName string, grafanaUrl string) payload.Payload {
+	id := uuid.New()
```
perhaps inline this variable, as it is defined quite far from its only usage?

atsu85 commented on server/initializer/initializer.go:
```diff
@@ -0,0 +1,91 @@
+	}
+}
+
+func createDashboardPayload(dashboardUid string, dashboardName string, grafanaUrl string) payload.Payload {
```
misleading name, as it creates panel, not dashboard?

tiithansen replied:
It creates payload to emulate dashboard visit.

atsu85 commented on server/initializer/initializer.go:
```diff
@@ -0,0 +1,91 @@
+	"time"
+)
+
+func InitializeMetricsForDashboards(api worker.Client, logger log.Logger, cache *cacher.Cacher) {
```
misleading name, as it doesn't initialize metrics, but instead adds a panel that starts sending metrics?

tiithansen replied:
It emulates visits to a dashboards to init metrics with 0 values. Username is used to compare if its emulated visit or not. If its emulated then session count (metric value) is not increased.

atsu85 commented on server/worker/api.go:
```diff
@@ -0,0 +1,141 @@
+	return body, nil
+}
+
+func (api *Client) prepare(method string, endpoint string, payload []byte) (*http.Request, error) {
```
perhaps rename to `prepareRequest`, to make its usage readable?

tiithansen replied:
I agree, move payload adding logic out from here also, essentially its helper to set some headers (including auth)

tiithansen commented on server/worker/api.go:
```diff
@@ -81,3 +124,18 @@ func (api *Client) Post(endpoint string, payload []byte) ([]byte, error) {
+	}
+
+	req.Header.Add("Authorization", bearer)
+	req.Header.Add("Content-Type", "application/json")
```
Remove, keep this as universal for any HTTP method. GET requests do not have body so `Content-Type` is not needed here.

Commit:
auto-analytics-adder - cr changes

simplify and increase readability

Commit:
Merge branch 'feature/auto-analytics-adder' into feature/init-dashboards-metrics

# Conflicts:
#	server/worker/api.go
#	server/worker/worker.go

Commit:
init-dashboards-metrics - Cr change - inline new uuid generation for better readability

Commit:
auto-analytics-adder - fix setting content type

Commit:
Merge branch 'feature/auto-analytics-adder' into feature/init-dashboards-metrics

Commit:
auto-analytics-adder - Change handler endpoint

Commit:
Merge branch 'feature/auto-analytics-adder' into feature/init-dashboards-metrics

Commit:
auto-analytics-adder - revert invalid change

Commit:
Merge branch 'feature/auto-analytics-adder' into feature/init-dashboards-metrics

Commit:
auto-analytics-adder - cr changes

Move filter check outside of updateDashboards func. Making function
signature correct, update might not of happened at all

increase request failed error information amount for better
understanding why the request failed

Commit:
Merge branch 'feature/auto-analytics-adder' into feature/init-dashboards-metrics

Commit:
init-dashboards-metrics - remove refactored method
